live/lib/auth_app/accounts.ex

```diff
@@ -235,119 +196,116 @@ defmodule AuthApp.Accounts do
   end
 
   @doc """
-  Deletes the signed token with the given context.
-  """
-  def delete_user_session_token(token) do
-    Repo.delete_all(UserToken.by_token_and_context_query(token, "session"))
-    :ok
-  end
-
-  ## Confirmation
-
-  @doc ~S"""
-  Delivers the confirmation email instructions to the given user.
-
-  ## Examples
-
-      iex> deliver_user_confirmation_instructions(user, &url(~p"/users/confirm/#{&1}"))
-      {:ok, %{to: ..., body: ...}}
-
-      iex> deliver_user_confirmation_instructions(confirmed_user, &url(~p"/users/confirm/#{&1}"))
-      {:error, :already_confirmed}
-
+  Gets the user with the given magic link token.
   """
-  def deliver_user_confirmation_instructions(%User{} = user, confirmation_url_fun)
-      when is_function(confirmation_url_fun, 1) do
-    if user.confirmed_at do
-      {:error, :already_confirmed}
+  def get_user_by_magic_link_token(token) do
+    with {:ok, query} <- UserToken.verify_magic_link_token_query(token),
+         {user, _token} <- Repo.one(query) do
+      user
     else
-      {encoded_token, user_token} = UserToken.build_email_token(user, "confirm")
-      Repo.insert!(user_token)
-      UserNotifier.deliver_confirmation_instructions(user, confirmation_url_fun.(encoded_token))
+      _ -> nil
     end
   end
 
   @doc """
-  Confirms a user by the given token.
+  Gets the user with the given signed token and deletes the used token.
+
+  Because of the way the magic link log in works, there are three cases to consider:
+
+    1. The user has not confirmed their email yet, no password is set.
+       In that case, the user gets confirmed and signed in.
+    2. The user has already confirmed their email. They are signed in.
+    3. The user has not confirmed their email yet, but a password is set.
+
+  The third case cannot happen with the default implementation, but it may happen
+  when adapting the code to a different use case, e.g. allowing to register with a password.
 
-  If the token matches, the user account is marked as confirmed
-  and the token is deleted.
+  This situation poses a security problem: an attacker may register with an email they
+  don't actually control which can lead to a session fixation attack:
+  The attacker can wait for the actual user to register the app. When they log in by magic link,
+  the account would get confirmed and the attacker would gain access by knowing the password
+  they set initially. To prevent this, the default implementation raises an error when trying
+  to use a magic link with a password set and unconfirmed email.
   """
-  def confirm_user(token) do
-    with {:ok, query} <- UserToken.verify_email_token_query(token, "confirm"),
-         %User{} = user <- Repo.one(query),
-         {:ok, %{user: user}} <- Repo.transaction(confirm_user_multi(user)) do
-      {:ok, user}
-    else
-      _ -> :error
+  def magic_link_login(token) do
```

Comment:
We need to have `user` in the function name. Suggestion:

```suggestion
  def login_user_by_magic_link(token) do
```

Here is my suggestion for the docs:

```
Logs the user in by magic link.

There are three cases to consider:
  
1. The user has already confirmed their email. They are logged in
   and the magic link is expired.
  
2. The user has not confirmed their email and no password is set.
   In this case, the user gets confirmed, logged in, and all tokens -
   including session ones - are expired. In theory, no other tokens
   exist but we delete all of them for best security practices.
   
3. The user has not confirmed their email but a password is set.
   This cannot happen in the default implementation but may be the
   source of security pitfalls. See the "Mixing magic link and password
   registration" section of `mix help phx.gen.auth`.
```

And we keep the bulk of the docs in the generator. We can update the error message to match accordingly.